### PR TITLE
[#213] [1/2 - Template] Apply to use Flow + get rid of custom UseCaseResult

### PR DESCRIPTION
**Pull request by luongvo**
Resolves https://github.com/nimblehq/android-templates/issues/213

## What happened 👀

- Apply to use [Flow](https://developer.android.com/kotlin/flow) to transfer the data from `Repository` through `UseCase` to `ViewModel`.
- Get rid of duplicating a new **try/catch** block in both `UseCase` and `Repository` when implementing https://github.com/nimblehq/android-templates/issues/212 as we can't access `retrofit objects` such as [HttpException](https://square.github.io/retrofit/2.x/retrofit/retrofit2/HttpException.html) in UseCase layer.
- Get rid of using a custom `UseCaseResult` to transfer data and exception from Usecase to ViewModel.
- Update all ViewModel logics and samples to adapt new structure.

## Insight 📝

- I don't think we should apply to use a custom `UseCaseResult` or even built-in [Result](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-result/), we can rely on kotlin.Flow to catch the error or collect data flow properly. 🤓 

## Proof Of Work 📹

The `template` app should work properly after migrating.

### CODE REVIEW

File: template/app/src/main/java/co/nimblehq/template/ui/screens/compose/HomeComposeViewModel.kt [hunk 1]
```diff
@@ -23,16 +21,17 @@ class HomeComposeViewModel @Inject constructor(
...
     init {
         execute {
-            when (val result = useCase.execute()) {
-                is UseCaseResult.Success -> {
-                    val uiModels = result.data.toUiModels()
-                    _uiModels.emit(uiModels)
-                }
-                is UseCaseResult.Error -> {
-                    val errorMessage = result.exception.message.orEmpty()
+            showLoading()
```
Comment (Wadeewee) [inline template/app/src/main/java/co/nimblehq/template/ui/screens/compose/HomeComposeViewModel.kt:24]: @luongvo Let's revert these changes. I saw you already updated in [this PR](https://github.com/nimblehq/android-templates/pull/292/files#diff-eb44200d9d6b03ba7a180a699c76534b141ad10297fb217d311a8fa4733a5a9d) So for this PR we can focus only the changes in `template.

File: template/data/src/main/java/co/nimblehq/template/data/repository/RepositoryImpl.kt [hunk 2]
```diff
@@ -4,10 +4,19 @@ import co.nimblehq.template.data.response.toModels
...
     private val apiService: ApiService
 ) : Repository {
 
-    override suspend fun getModels(): List<Model> = apiService.getResponses().toModels()
+    override fun getModels(): Flow<List<Model>> = flow {
+        try {
+            val result = apiService.getResponses().toModels()
+            emit(result)
+        } catch (e: Exception) {
+            throw e
```
Comment (Wadeewee) [inline template/data/src/main/java/co/nimblehq/template/data/repository/RepositoryImpl.kt:19]: No biggie, but better to name it more clearly.
```suggestion
        } catch (exception: Exception) {
            throw exception
```

File: template/domain/src/main/java/co/nimblehq/template/domain/usecase/UseCase.kt [hunk 3]
```diff
@@ -2,16 +2,12 @@ package co.nimblehq.template.domain.usecase
...
 
-    suspend fun execute(): UseCaseResult<List<Model>> {
-        return try {
-            val response = repository.getModels()
-            UseCaseResult.Success(response)
-        } catch (e: IllegalStateException) {
-            UseCaseResult.Error(e)
-        }
+    fun execute(): Flow<List<Model>> {
+        return repository.getModels()
```
Comment (Wadeewee) [inline template/domain/src/main/java/co/nimblehq/template/domain/usecase/UseCase.kt:11]: @luongvo Hmm, Do we still need `UseCase` if we don't handle any of the results(`success/error`) in `UseCase`? I worried that it would have a lot of logic inside `VM` instead 🤔

File: template/app/src/main/java/co/nimblehq/template/ui/screens/compose/HomeComposeViewModel.kt (hunk 1 above)
Comment (luongvo) [reply template/app/src/main/java/co/nimblehq/template/ui/screens/compose/HomeComposeViewModel.kt:24]: @Wadeewee Yes, but it's in the template 😂

File: template/domain/src/main/java/co/nimblehq/template/domain/usecase/UseCase.kt (hunk 3 above)
Comment (luongvo) [reply template/domain/src/main/java/co/nimblehq/template/domain/usecase/UseCase.kt:11]: @Wadeewee From what we can see from VM implementation, we have `catch` and `collect` to subscribe for success or error case already. We can leave the usecase back for some data preparation (request) or data consolidation (response) actions. 🤔

File: template/data/src/main/java/co/nimblehq/template/data/repository/RepositoryImpl.kt (hunk 2 above)
Comment (ryan-conway) [reply template/data/src/main/java/co/nimblehq/template/data/repository/RepositoryImpl.kt:19]: @luongvo Quick question, why did we wrap this in a try catch if we immediately throw the caught exception anyway? 🤔

File: template/data/src/test/java/co/nimblehq/template/data/repository/RepositoryTest.kt [hunk 4]
```diff
@@ -9,6 +9,7 @@ import io.kotest.matchers.shouldBe
 import io.mockk.coEvery
```
Comment (Wadeewee) [inline template/data/src/test/java/co/nimblehq/template/data/repository/RepositoryTest.kt:9]: Please remove `import io.kotest.assertions.throwables.shouldThrow`

File: template/data/src/main/java/co/nimblehq/template/data/repository/RepositoryImpl.kt (hunk 2 above)
Comment (luongvo) [reply template/data/src/main/java/co/nimblehq/template/data/repository/RepositoryImpl.kt:19]: @ryan-conway if we throw it, we need to catch it at usecase or VM level.
- Catching at usecase level: it's hard for us to check the exception's body such as `HttpException` which can not access from **domain** module. Please checkout more insights here https://github.com/nimblehq/android-templates/issues/213#issuecomment-1135375842. 👋 


- Catching at VM level: we need to have 1 more try/catch to wrap the Flow collector. It's a bad idea I think

```
try{
              useCase.execute()
                .catch {
                   // catch error
                }
                .collect { 
                    // get value
                }
}catch (e: Exception){
   // catch error
}
```

Comment (ryan-conway) [reply template/data/src/main/java/co/nimblehq/template/data/repository/RepositoryImpl.kt:19]: @luongvo Ah, so this is just preparation for the custom error mapping, right? 😄

Comment (luongvo) [reply template/data/src/main/java/co/nimblehq/template/data/repository/RepositoryImpl.kt:19]: @ryan-conway exactly, we will have something like this

image

File: template/domain/src/main/java/co/nimblehq/template/domain/usecase/UseCase.kt (hunk 3 above)
Comment (Wadeewee) [reply template/domain/src/main/java/co/nimblehq/template/domain/usecase/UseCase.kt:11]: > From what we can see from VM implementation, we have catch and collect to subscribe for success or error case already. It's very clear.

I understand this point: ☝️ @luongvo, but my concern is when handling the chain calls because the previous way with `UseCaseResult` in `UseCase`, we can inject another `UseCase,` So whenever it's `success/error` and we need to call something from another UseCase to get the result, Let's say we want to call some `API` when the error occurs So, by this way we can execute the next `UseCase` in current UseCase instead of filter in `VM` 🙏  It seems like using `UseCaseResult` it's more flexible, IMO🤔

Comment (luongvo) [reply template/domain/src/main/java/co/nimblehq/template/domain/usecase/UseCase.kt:11]: > So, by this way we can execute the next UseCase in current UseCase instead of filter in VM 🙏 It seems like using UseCaseResult it's more flexible, IMO🤔

@Wadeewee I think you're missing and remember how `flatMap` benefits us when working with RxJava. So switching to Flow, we have something called `flatMapConcat` to map multiple flows in a chain. Please checkout more sample here https://github.com/nimblehq/android-templates/issues/213#issuecomment-1135375842

Do you think which one is better, the left or the right side? 🤓 

image

Comment (Wadeewee) [reply template/domain/src/main/java/co/nimblehq/template/domain/usecase/UseCase.kt:11]: > Do you think which one is better, the left or the right side? 🤓

@luongvo If we handle the `Chain` call in `VM` like your example.  Surely, the right side without `UseCaseResult` is better, but Actually, we can handle it in `UseCase` instead of `VM,` So in `VM,` we will handle it only 
- if success -> emit result value to view or do something but not `execute another usecase.`
- If error -> emit error to view or do something but not `execute another usecase.` 

Currently, our template doesn't have complex logic, So I think it's good to go without `UseCaseResult,` but TBH, I'm not sure for the `Client` project that might have complex logic.

**Comment by chornerman** [comment]
@luongvo I think there are some duplicate words in the description 😄 
Screen Shot 2565-09-01 at 12 38 39

**Review by ryan-conway — COMMENTED** [review]
LGTM aside from the comments [here](https://github.com/nimblehq/android-templates/pull/291/files#r959131964) and [here](https://github.com/nimblehq/android-templates/pull/291/files#r959141388) 👍🏻

File: template/data/src/main/java/co/nimblehq/template/data/repository/RepositoryImpl.kt (hunk 2 above)
Comment (sleepylee) [reply template/data/src/main/java/co/nimblehq/template/data/repository/RepositoryImpl.kt:19]: better add TODO

**Review by chornerman — COMMENTED** [review]
LGTM aside from Ryan's comment 😄

File: template/domain/src/main/java/co/nimblehq/template/domain/usecase/UseCase.kt (hunk 3 above)
Comment (luongvo) [reply template/domain/src/main/java/co/nimblehq/template/domain/usecase/UseCase.kt:11]: @Wadeewee A usecase should do a single job only and it should not contain a chain logic there. We should keep the use case calls chain in VM layer instead.

File: template/data/src/test/java/co/nimblehq/template/data/repository/RepositoryTest.kt (hunk 4 above)
Comment (luongvo) [reply template/data/src/test/java/co/nimblehq/template/data/repository/RepositoryTest.kt:9]: @Wadeewee Fixed in https://github.com/nimblehq/android-templates/pull/292/commits/0d2563c9390e5a2bdfc1ae122af7d966f7093594

File: template/data/src/main/java/co/nimblehq/template/data/repository/RepositoryImpl.kt (hunk 2 above)
Comment (luongvo) [reply template/data/src/main/java/co/nimblehq/template/data/repository/RepositoryImpl.kt:19]: @Wadeewee Fixed in https://github.com/nimblehq/android-templates/pull/292/commits/0d2563c9390e5a2bdfc1ae122af7d966f7093594

**Review by chornerman — APPROVED** [review]
LGTM 🚀

**Comment by lydiasama** [comment]
Do we need to update the sample projects after merging this to template? @ryan-conway @luongvo

**Review by lydiasama — APPROVED** [review]

**Comment by ryan-conway** [comment]
@lydiasama Yes, that will be done in #292  😄

**Review by ryan-conway — APPROVED** [review]
👌🏻

**Review by Wadeewee — APPROVED** [review]
